docs(rfc): add declarative configuration rfc (#322)
getoutreach/devbase: rfcs/1-declarative-configuration.md

# Declarative Configuration

## Table of Contents

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [Implementation Details](#implementation-details)
  - [Configuration example](#examples)
- [Test Plan](#testrelease-plan)
<!-- /toc -->

## Summary

This proposal replaces environment variable based configuration being used in Make+Magefiles in favor of using declarative yaml based configuration.

## Motiviation

> As we've been writing `devbase`, and other tooling, we've identified that trying to codify all options for tooling to environment variables, bespoke configuration files, and other obscure means has made it difficult for us to provide options and maintain a common set of toolign.
> 
> This, combined with a lack of documentation into all of these options, has laregely made `devbase` unconsumable outside of bootstrap/stencil. Thinking of `devbase` as it's own framework to build software makes it more usable in non-stencil projects (like we do already today w/ submodules)

Reviewer: ```suggestion
As we've been writing `devbase` and other tooling we've identified that trying to codify all options for tooling to environment variables, bespoke configuration files, and other obscure means has made it difficult for us to provide options and maintain a common set of tooling. This, combined with a lack of documentation into all of these options, has largely made `devbase` inconsumable outside of bootstrap/stencil. Thinking of `devbase` as its own framework to build software makes it more usable in non-stencil projects (like we do already today w/ submodules).
```

Reviewer: im someone who stumbled upon this open source repository (i do not work for outreach)

> (like we do already today w/ submodules)

wat?

### Goals

>  - Documentation for all options in configuration file(s)

Reviewer: I keep having an issue with how vague the word options is

Author: It's somewhat intentionally vague as it's different dependent on the target, I could all it configuration instead? However that seems equally vague.

Reviewer 2: Maybe this should be comments for all options in config files. I think keeping this info as close to the config as possible will have a big benefit.

Author: PTAL, I updated this to be more clear about what I'm envisioning

Reviewer 2: Big fan of the new direction of levering Go docs.

 - Standard location for all configuration for tooling provided by devbase, et. al
 - Documentation mapping 1:1 with Magefile target, with framework to consume/declare it
 - Documentation on how to write Magefile targets, if the framework docs aren't self-explanatory enough

>  - Support for usage outside of stencil-base, et. al

Reviewer 2: To add to this, if we plan on making this usable outside of stencil, devbase needs its own init script (which can be stencil based). That script can take care of dropping config files with default configs in the right folder and adding other things.

>  - Support for usage outside of stencil-base, et. al

Author: I want defaults to be in-code, I don't want configuration to be required unless it's needed.

`docker.yaml` is a good example, the entirety of the configuration works OOTB, you just have to provide a key for the docker image name. If you stray from those defaults, you can override them.

>  - Support for usage outside of stencil-base, et. al

Reviewer 2: Totally makes sense. If we have the documentation available externally as Go docs, this approach works perfectly.

>  - Not breaking, while we could do a lot more if we did a breaking release, breaking _all_ of the existing releasing tooling isn't optimal.

Reviewer: ```suggestion
 - Not breaking, while we could do a lot more if we did a breaking release, breaking _all_ of the existing releasing tooling isn't optimal
```
when u make a list either use end of sentence punctuation or dont but don't mix the two

### Non-goals

>  - Expose new functionality, this would be nice to have but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.

Reviewer: ```suggestion
 - Expose new functionality. This would be nice to have, but would balloon the amount of work. We're targeting 1:1 compatibility with features/configuration already exposed today, with nice-to-haves being limited.
```

>  - Migrating `Makefile` to `Magefile`, while it'd be nice to migrate all of them over, that's also an amount of work to do. While moving to `Magefile` would be good, we shouldn't try to move everything to Go at the same time. (TL;DR: Wrap shell were needed)

Reviewer: ```suggestion
 - Migrating `Makefile` to `Magefile`, while it'd be nice to migrate all of them over, that's also an amount of work to do. While moving to `Magefile` would be good, we shouldn't try to move everything to Go at the same time (TL;DR: Wrap shell were needed).
```

punctuation goes after parenthesis at the end of a sentence

## Proposal

The high level proposal here is generally covered in the [Summary](#summary) and [Moviation](#motiviation) sections.

### Implementation Details

#### Repository Configuration

Configuration would live in the root of repository consuming `devbase`, in `.devbase`.

> Each configuration file would match the relevant `mage <target>` target, e.g.

Reviewer 2: If we are going to continue to support `make`, for the time being, this should state `make/make <target>`.

```bash
$ ls .devbase
build.yaml
e2e.yaml
docker.yaml
```

> **Note**: For shared configuration, e.g. `docker` (which could be shared between `docker-build` and `docker-push`) that's also allowed and can become a unified target.

Reviewer: You should come up with a paradigm here. Let's take the same example with `docker-build` and `docker-push` sharing configuration in `docker.yaml`

Maybe the paradigm is this - for these targets with shared config they need to all be postfixed with `docker_` and you write shared config in the root of the yaml document, with config pertaining to that specific target under a key with the same name as what follows the prefix.

So what you'd have is actually `docker_build` and `docker_push` with a `docker.yaml` file that looks like this:
```yaml
i_am: "shared config"
here_is_more: "shared config"
build:
  more: "config"
push:
  moar: "config"
```

Reviewer: the reason i did underscore instead of hyphen is because then u can still use hypen if need multiple words after the postfixed part

Author: Talked offline, updated this note to be more clear

Reviewer 2: We could also nest this arbitrarily. e.g. every config file has a nested structure.

For example, `docker.yaml` would have only the config for all docker targets:
```yaml
docker:
  i-am-shared-for-docker-targets: "config"
```

Then `docker-build.yaml` would have the config for build:
```yaml
docker:
  build:
    i-am-only-used-in-build: "config"
```

That could also allow to easily overriding a config from `docker.yaml` inside `docker-build.yaml`.

> **Note**: For shared configuration, e.g. `docker` (which could be shared between `docker-build` and `docker-push`) that's also allowed and can become a unified target.

Reviewer 2: I could see having multiple ways to share configs quickly getting out of had. One way to handle this would be to have a single shared variable file and have a hierarchy where all targets get the shared file, but can supersede these options in their specific config files.

This might result in a big shared variable file, but it keeps things simple until we need the complexity.

Author: PTAL at the new wording that expresses what I meant to say here

#### Magefile Targets Implementation Details

> Each magefile target that consumes this configuration will live inside of the `devbase` repository in the `targets/<targetName>` package, which will be pulled in by `root/mage.go`

Reviewer: ```suggestion
Each magefile target that consumes this configuration will live inside of the `devbase` repository in the `targets/<targetName>` package, which will be pulled in by `root/mage.go`.
```

> Each magefile target that consumes this configuration will live inside of the `devbase` repository in the `targets/<targetName>` package, which will be pulled in by `root/mage.go`
> 
> The reasoning for a package per target is to make it easier to test these in isolation, as well as (generated) documentation to be localized to the packages (targets) themselves.
> 
> A library will be provided at `pkg/targets`, which will provide functionality to read configuration, and other shared functions between targets.

Reviewer: ```suggestion
Each magefile target that consumes this configuration will live inside of the `devbase` repository in the `targets/<targetName>` package, which will be pulled in by `root/mage.go`. The reasoning for a package per target is to make it easier to test these in isolation, as well as (generated) documentation to be localized to the packages (targets) themselves. A library will be provided at `pkg/targets`, which will provide functionality to read configuration and other shared functions between targets.
```

u have three related sentences here, each acting as their own paragraph

#### Magefile Targets Available Day One

The following targets will be available/configurable day one:

 - `build`
 - `debug`
 - `release`
 - `test` (includes: `coverage` as `--coverage`, `benchmark` as `--bench`)
 - `lint`
 - `fmt`
 - `e2e`
 - `docker` (includes: `build`, `publish`)

The rest will still be available for compat, but unconfigurable, e.g. `gobuild` will still build go but will not work for non-go repos.

See [Examples](#examples) for an idea of what options will be available/an idea of some of the steps.

#### Testing Framework

> Testing would be done via Go Testing, enabling us to easily write unit tests. Integration tests would, unforuantely, still need to be figured out as they rely on being ran in macOS/Linux in a predictable fashion

Author: Maybe we should consider writing this to run them all in containers, or something else that'd be more testable...

  ldflags:
    github.com/getoutreach/gobox/pkg/app.Version: '{{ .Version }}'
```

Or `docker.yaml`, that already [exists today](https://github.com/getoutreach/devbase#building-docker-images):

```yaml
# Corresponds to the directory in `deployments/` and is used, by default, as the image name (see special case below).
# If this is equal to "appName" then the build context/image name is changed. See docs.
image_name:
  # Override the build context for the image. Defaults to '.' if image_name == appName, otherwise ./deployments/<image_name>
  buildContext: .
  # pushTo is a different image registry to push to, defaults to:
  # (box: .devenv.imageRegistry)/$image (or, if not == appName, $appname/$image)
  pushTo: docker.com/stable/athens
  # extra secrets to expose to the builder, defaults to NPM_TOKEN being exposed
  # See: https://docs.docker.com/develop/develop-images/build_enhancements/#new-docker-build-secret-information
  secrets:
    - id=mySecret,env=MY_SECRET_ENV_VARIABLE
  # Platforms to build this image for, defaults to linux/amd64,linux/arm64
  # See: https://github.com/docker/buildx#building-multi-platform-images
  platforms:
    - linux/amd64
 ```

## Test/Release Plan

> The test plan for this is to ensure that each target has unit, and where possible, integration tests. Releasing would follow the standard DT team releasing process, which is to be released in the next release window, while being put on the `rc` channel to be consumed by Outreach internally first.

Reviewer: ```suggestion
The test plan for this is to ensure that each target has unit and, where possible, integration tests. Releasing would follow the standard DT team releasing process, which is to be released in the next release window, while being put on the `rc` channel to be consumed by Outreach internally first.
```
